### PR TITLE
Resolve precision loss in functions that call `_removeCollateral()` and withdraw collateral to the caller

### PR DESCRIPTION
**Pull request by nicholaspai**
This PR enforces that `withdraw`, `redeem`, `settleExpired` and `withdrawLiquidation` all transfer from the contract the exact amount of collateral that the position's collateral is decreased by. The tradeoff is that the user might receive 1e-18 less collateral than they expected to receive.

The bulk of this PR is additional unit tests to make sure that contracts are dealing with precision loss in the expected way.

More specifics below:
___
In `withdraw()`, `redeem()`, `settleExpired()`, and `withdrawLiquidation()`, collateral is transferred out of the contract via `safeTransfer(amount)` while the internal counter `rawTotalPositionCollateral` or `rawLiquidationCollateral` is decreased. 

The ideal is to always maintain the invariant that `(rawCollateral * cumulativeFeeMultiplier)` is less than or equal to `collateral.balanceOf(EMP.address) = collateralOwned`. If this were not true, then the contract could lock by trying to withdraw more collateral than it actually owns.

This PR resolves one class of precision errors demonstrated in #1037.

Any function that calls `_removeCollateral` followed by a `safeTransfer(caller, amount)` will cause `totalPositionCollateral` to **decrease** by less than is transferred out, because `_removeCollateral` produces precision loss. (Conversely, any function that calls `_addCollateral` followed by `safeTransferFrom(caller, this, amount)` causes `totalPositionCollateral` to **increase** by less than is transferred in, which is OK).

Signed-off-by: Nick Pai <npai.nyc@gmail.com>

### CODE REVIEW

**Comment by nicholaspai** [comment]
I debated whether `createLiquidation` needed a change, but this method is subtlly different than the above ones that call `_removeCollateral()` followed by `collateral.safeTransfer():

- `createLiquidation` first calls `_reduceSponsorPosition()` which decreases `totalPositionCollateral` by the same amount that it adds to it via a subsequent `_addCollateral()` . Therefore, there is no precision lost between the collateral being "removed" from the position and the collateral being "added" to the liquidation.
- Moreover, this method doesn't actually transfer any collateral currency out of the contract, so there is no precision lost between `totalPositionCollateral` and `collateral`

File: core/contracts/financial-templates/implementation/FeePayer.sol [hunk 1]
```diff
@@ -177,13 +177,26 @@ abstract contract FeePayer is Testable {
...
         rawCollateral.rawValue = rawCollateral.sub(adjustedCollateral).rawValue;
+        removedCollateral = initialBalance.sub(_getCollateral(rawCollateral));
     }
 
+    // Increase rawCollateral by a fee-adjusted collateralToRemove amount. Fee adjustment scales up collateralToRemove
+    // by dividing it by cumulativeFeeMutliplier. There is potential for this quotient to be floored, therefore rawCollateral
+    // is increased by less than expected. Because this method is usually called in conjunction with an actual addition of collateral
+    // into this contract, there is no need to return the fee-adjusted amount that the rawCollateral is increased by because the contract
+    // will have surplus collateral in case an error is introduced.
     function _addCollateral(FixedPoint.Unsigned storage rawCollateral, FixedPoint.Unsigned memory collateralToAdd)
```
Comment (mrice32) [inline core/contracts/financial-templates/implementation/FeePayer.sol:200]: Drive-by-comment: why not make this symmetric and do the same thing when we take collateral *in* to reduce rounding there too?

File: core/scripts/PrecisionErrors.js [hunk 2]
```diff
@@ -502,23 +504,37 @@ async function runExport() {
...
+  //   `totalPositionCollateral` is 0.1e-18 higher than `collateral`. Each withdraw increases this error, which only becomes visible after the tenth withdrawal.
+  // - Let's look at the second `withdraw()`
+  //        before the second `withdraw()`, `totalPositionCollateral = 0.299999999999999999`
+  //        after the second `withdraw()`, `totalPositionCollateral = 0.999999999999999994 * 0.3 = 0.2999999999999999982` which truncates the last decimal
+  // - Again, this is as expected: `totalPositionCollateral` is decreased by 1e-18, so we can transfer 1e-18 to the user
+  // - In fact, the error does not emerge until the tenth `withdraw`. After the ninth withdraw, `rawCollateral = 0.999999999999999973` (1e-18 + 9 * 3e-18),
+  //        before the tenth `withdraw()`, `totalPositionCollateral = 0.999999999999999973 * 0.3 = 0.2999999999999999919`
+  //        after the tenth `withdraw()`, `totalPositionCollateral = 0.9999999999999999970 * 0.3 = 0.2999999999999999910` which truncates the last decimal
+  // - So, this is unexpected: `totalPositionCollateral` will show no decrease. After each withdrawal, `totalPositionCollateral` 0.9e-18 instead of the full
```
Comment (chrismaree) [inline core/scripts/PrecisionErrors.js:531]: This is really interesting, great analisis!

File: core/scripts/PrecisionErrors.js [hunk 3]
```diff
@@ -642,20 +662,21 @@ async function runExport() {
...
+  // Example of (1):
   // - M = cumulativeFeeMultiplier = 1
   // - RC = rawCollateral = 9
   // - T = tokensOutstanding = 9
   // - S = amountToRedeem = 8
   // - (S / T) = 0.888...repeating and gets floored to 0.888...88
   // - (S / T) * (RC * M) = 7.999...92, instead of 8
   // - So, the user burns 8 synthetic but receives 7.999...92 collateral, representing 8e-18 less than they were expecting.
+  // Error Amount:
+  // - (-1e-17) in `c` if `(S / T)` has -1e-18 error and is then multiplied by C's least significant decimal
```
Comment (chrismaree) [inline core/scripts/PrecisionErrors.js:677]: Not directly related to your PR but this is where the 12 orders of magnitude error increase with USDC will come in as this error compunds by C's least significant decimal.

File: core/scripts/PrecisionErrors.js [hunk 4]
```diff
@@ -642,20 +662,21 @@ async function runExport() {
...
 
-  console.group("Precision loss in proportion of collateral redeemed via redeem()");
+  console.group("Precision loss in % of collateral redeemed via redeem()");
```
Comment (chrismaree) [inline core/scripts/PrecisionErrors.js:679]: Great to express it as a % of the total collateral redeemed, as you do in you comments above.

File: core/test/financial-templates/PricelessPositionManager.js [hunk 5]
```diff
@@ -940,23 +964,37 @@ contract("PricelessPositionManager", function(accounts) {
...
+    // The value of the remaining synthetic tokens = 10 * 1.2 = 12.
+    // So, we will attempt to withdraw (12 + 4) tokens from the contract.
+    // We need to decrease `rawCollateral` by `16 / 0.966666666666666666 ~= 16.5`
+    // which gets truncated to 16.
+    // Recall previously that rawCollateral was last set to 18, so `totalPositionCollateral = (18-16) * 0.966666666666666666 ~= 1.97`
+    // which gets truncated to 1.
+    // The previous totalPositionCollateral was 17, so we will withdraw (17-1) = 16 tokens instead of the 17 as the user expected.
+    assert.equal((await pricelessPositionManager.totalPositionCollateral()).toString(), "1");
+    assert.equal((await pricelessPositionManager.rawTotalPositionCollateral()).toString(), "2");
+    const expectedSponsorCollateralSynthetic = toBN("11");
```
Comment (chrismaree) [inline core/test/financial-templates/PricelessPositionManager.js:980]: This is excellent and your comments perfectly show how this PR's change to under credit in favour of rounding results in this single wei less in the transfer.

File: core/contracts/financial-templates/implementation/FeePayer.sol (hunk 1 above)
Comment (nicholaspai) [reply core/contracts/financial-templates/implementation/FeePayer.sol:200]: So I definitely debated this but the tradeoff for me was practicality versus symmetry.

Symmetrically it makes sense to do the same thing when taking collateral in (`safeTransferFrom`) and calling `_addCollateral`.

Practically, we are OK if there is more collateral transferred in then added to `totalPositionCollateral`.

What do you think?

File: core/scripts/PrecisionErrors.js (hunk 3 above)
Comment (nicholaspai) [reply core/scripts/PrecisionErrors.js:677]: Yep I think we can't avoid the precision loss even with currencies that have fewer decimals. In fact the precision will be amplified

File: core/contracts/financial-templates/implementation/FeePayer.sol (hunk 1 above)
Comment (mrice32) [reply core/contracts/financial-templates/implementation/FeePayer.sol:200]: While I totally agree that practically, it doesn't help much, I think the consistency and symmetry makes the code easier to read and understand. Obviously, if this adds a ton more complexity, then maybe it's not worth it, but IMO, it would be better just to do it in both directions so it's clearer that these operations are exactly opposite of one another.

Comment (nicholaspai) [reply core/contracts/financial-templates/implementation/FeePayer.sol:200]: I've added the return value to `_addCollateral` but I'm still undecided about whether we should actually use that return value as input in `safeTransferFrom` for deposits/creates.

The problem with this is that sometimes the `totalPositionCollateral` is increased by 0 due to precision loss. Normally we'd still transfer 1 collateral from the user, but if we make a change here then its possible for deposits to transfer 0. However, if the method does not revert, then the `rawTotalPositionCollateral` will increase, but it will not reflect an increase in the `totalPositionCollateral`. It's a really odd case, but maybe not a bug.

File: core/contracts/financial-templates/implementation/PricelessPositionManager.sol [hunk 6]
```diff
@@ -196,12 +196,15 @@ contract PricelessPositionManager is FeePayer, AdministrateeInterface {
...
         _removeCollateral(positionData.rawCollateral, collateralAmount);
         require(_checkPositionCollateralization(positionData));
-        _removeCollateral(rawTotalPositionCollateral, collateralAmount);
+        FixedPoint.Unsigned memory amountWithdrawn = _removeCollateral(rawTotalPositionCollateral, collateralAmount);
 
         // Move collateral currency from contract to sender.
-        collateralCurrency.safeTransfer(msg.sender, collateralAmount.rawValue);
+        // Note that we move the amount of collateral that is decreased from rawCollateral (inclusive of fees)
+        // instead of the user requested amount. This eliminates precision loss that could occur
+        // where the user withdraws more collateral than rawCollateral is decremented by.
```
Comment (mrice32) [inline core/contracts/financial-templates/implementation/PricelessPositionManager.sol:204]: nit: can you add a note about why it only matters for the global collateral and not the position's collateral? I think our reasoning is that we need to ensure our bookkeeping around the global collateral is always <= the true amount of collateral in the contract to avoid reverts on withdrawals.

File: core/contracts/financial-templates/implementation/PricelessPositionManager.sol [hunk 7]
```diff
@@ -490,15 +498,19 @@ contract PricelessPositionManager is FeePayer, AdministrateeInterface {
...
-        rawTotalPositionCollateral = rawTotalPositionCollateral.sub(positionToLiquidate.rawCollateral);
+        FixedPoint.Unsigned memory remainingRawCollateral = positionToLiquidate.rawCollateral;
+        rawTotalPositionCollateral = rawTotalPositionCollateral.sub(remainingRawCollateral);
         totalTokensOutstanding = totalTokensOutstanding.sub(positionToLiquidate.tokensOutstanding);
 
         // Reset the sponsors position to have zero outstanding and collateral.
         delete positions[sponsor];
+
+        // Return fee-adjusted amount of collateral deleted from position.
+        return _getCollateral(remainingRawCollateral);
```
Comment (mrice32) [inline core/contracts/financial-templates/implementation/PricelessPositionManager.sol:513]: Maybe I'm missing something, but shouldn't we compute the difference between the global collateral before and after the subtraction?

Comment (nicholaspai) [reply core/contracts/financial-templates/implementation/PricelessPositionManager.sol:533]: in this case all of the position's collateral is being withdrawn from the global collateral, but I see a point in that there might rounding errors here. I'll adjust

Comment (nicholaspai) [reply core/contracts/financial-templates/implementation/PricelessPositionManager.sol:533]: done

File: core/contracts/financial-templates/implementation/PricelessPositionManager.sol (hunk 6 above)
Comment (nicholaspai) [reply core/contracts/financial-templates/implementation/PricelessPositionManager.sol:209]: done

**Comment by nicholaspai** [comment]
This still needs a test for withdrawLiquidation, hence why it’s in draft mode
